Improve Native Image Inlining

roberttoyonaga opened the pull request:
# Summary
This PR improves Native Image inlining. The new inlining involves 3 steps:
1. Trivial inlining
2. Non-Trivial inlining
3. Single callsite inlining

### Trivial Inlining
This stage remains mostly the same as the original default inliner. It is a quick operation that inlines methods under a certain size threshold. Each callee method is assessed independently against the threshold. I have updated the method size computation to use `Node.estimatedNodeSize()` rather than a simple raw Graal-IR node count. 

Inlining happens in rounds. During each round, methods that have been marked trivial (below size threshold) in the previous round are inlined. At the end of the round, each method in the universe is assessed to determine whether it has become newly trivial due to the inlining completed over the course of the round. If so, it is marked for inlining the next round. 

Note that `MaxNodesInTrivialMethod` and `MaxNodesInTrivialLeafMethod` have been replaced with `MaxTrivialMethodSize` and `MaxTrivialLeafMethodSize`.  The defaults are `-H:MaxTrivialMethodSize=55` and `-H:MaxTrivialLeafMethodSize=128`

### Non-Trivial Inlining
This is a new stage that focuses on inlining methods that are predicted to yield good optimization benefit due to specialization from inlining. This is determined through "trialing" the inlining before committing it. When inlining of a method is trialed, the successful canonicalizations (basic compiler optimizations) are recorded and counted as "benefit". The "benefit" from inlining and the cost (size of method) are used to decide whether the inlining trial should be committed or rolled back.  

Similar to the Trivial Inlining stage, Non-Trivial inlining assesses each callee independently, and inlining happens in rounds. 

Similar to the Trivial Inlining stage, this Non-Trivial Inlining allows for nested (recursive) inlining. However, only one level of inlining happens each round. This is because inlining trials are committed from the leaves backward toward the root (similar to the Inlining Before Analysis stage). An inlining decision for callees will be reached before their caller. If we inline multiple levels deep in a singe round, we may waste effort trialing Nth-level callees when the 1st-level callee inlining trial will later fail. 

### Single Callsite Inlining
This is a new stage that focuses on methods that have only 1 callsite.  Inlining of single callsite methods should not duplicate code area. The number of callsites for each method is counted during Non-Trivial Inlining stage. This stage is very similar to the Trivial Inlining stage, but the criteria for inlining is not based on size.  This stage is saved for last to avoid code area duplication, please see the Javadoc for `inlineSingleCallsiteMethods` for more details.

roberttoyonaga committed:
stage one with comments

roberttoyonaga committed:
use root original size in threshold

roberttoyonaga committed:
use a limit on the number of inline attempts based on original callee count

roberttoyonaga committed:
weight specific canon differently. Multiple callsites. Fix sizebeforeInlining bug

roberttoyonaga committed:
no invokes is better, flip ignored flag on changed

roberttoyonaga committed:
callsiteCount=1, then inline. Update 2nd lvl callee callsiteCounts when inlining targets

roberttoyonaga committed:
remove prioritization

roberttoyonaga committed:
clean up: remove canImproveStamps

roberttoyonaga committed:
clean up: remove commented out code

roberttoyonaga committed:
clean up: remove IGV dumping

roberttoyonaga committed:
clean up: remove guiding *** comments

roberttoyonaga committed:
clean up: remove old size calculations.

roberttoyonaga committed:
clean up: refactor killControlFlowNodes, clean up comments

roberttoyonaga committed:
clean up: remove unused variables and comments.

roberttoyonaga committed:
clean up: factor out NonTrivialInliningGraphDecoder. Other refactoring

roberttoyonaga committed:
fix recursion and add TrivialInlining and NonTrivialInlining after SingleCallsiteInlining

roberttoyonaga committed:
clean up: refactor and minor simplify

roberttoyonaga committed:
add inlining potential pre-screening to speed up build

roberttoyonaga committed:
clean up: style, cosmetics

roberttoyonaga committed:
improve TrivialInliningPhase to use estimated size

roberttoyonaga committed:
clean up: collapse InlinePhases into a single class

roberttoyonaga committed:
clean up: refactor canInline

roberttoyonaga committed:
clean up: prints, comments and refactor

minor rebase fix

clean up

roberttoyonaga committed:
style

roberttoyonaga committed:
prevent recursive inlining due to annotations

roberttoyonaga committed:
style and small fix for layers

roberttoyonaga committed:
clean up: simplify inlineSingleCallsiteMethods and makeSingleCallsiteInlineDecision

comments

roberttoyonaga committed:
Add size fallbakc for single callsite method inlining. Some refactoring.

roberttoyonaga committed:
minor tweaks to weights based on experimentation

roberttoyonaga committed:
clean up: tighten access modifiers, spelling, grammar

zakkak commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/SimplifyingGraphDecoder.java:
```diff
@@ -287,6 +288,7 @@ protected boolean earlyCanonicalization(MethodScope methodScope, LoopScope loopS
             methodScope.reader.setByteIndex(successorsByteIndex + (IfNode.SUCCESSOR_EDGES_COUNT * methodScope.orderIdWidth));
 
             removeSplit(methodScope, loopScope, ifNode, survivingOrderId);
+            methodScope.benefit += 55;
```
Please define such constants as constant fields and add a comment explaining why `55`, `35`, etc.

roberttoyonaga replied:
I've changed these weights to be [constant fields and added a new comment ](https://github.com/roberttoyonaga/graal/pull/3/files#diff-f138ede480c3d7e8b9136e80a810ff20be44c74f34a0a2f2ef36ba87c0633c9dR70-R81)generally explaining my reasoning. However, I arrived at the specific weights experimentally, so I primarily only have [this spreadsheet](https://docs.google.com/spreadsheets/d/1XLzMnGXZlWyLJiolYbK_iNRgCt1ms2-gQbRoq3vPiTY/edit?gid=391151018#gid=391151018) to justify them.  This is one of the things I am worried about when it comes time to push this upstream.

roberttoyonaga replied:
That being said, these weights only get used by the Non-Trivial stage. I plan to adopt your suggestion and break this into 3 PRs. We can see how the Trivial and SingleCallsite stages are received upstream first and adapt accordingly.

galderz commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/GraphDecoder.java:
```diff
@@ -140,6 +140,7 @@ protected class MethodScope {
          * inlining log. {@code null} if the inlining log is not being decoded.
          */
         public InliningLogCodec.InliningLogDecoder inliningLogDecoder;
+        public int benefit = 0;
```
Can you add javadoc on what benefit number represents and how it works? There are several different values used to update it, e.g. 1, 20, 35...  but it's not clear what each of those mean and why each have specific values.

roberttoyonaga replied:
Yes that's a good point. These hardcoded values stand out a lot.  I've added some new comments [here](https://github.com/roberttoyonaga/graal/pull/3/files#diff-f138ede480c3d7e8b9136e80a810ff20be44c74f34a0a2f2ef36ba87c0633c9dR70-R81) explaining generally how the values were chosen.

The numbers are unit-less values that get added to the benefit sum. Later, in [`canInline`](https://github.com/roberttoyonaga/graal/pull/3/files#diff-139b361e75fec4c2eb329444c15758e8af020843e7a24b2cbe9d67f88ef7a297R127), a higher benefit sum value will increase the likelihood of the inlining trial being committed. I'm not sure how to comment a theoretical justification for each weight specifically because the values are empirically determined (see[ parameter space experiment results](https://docs.google.com/spreadsheets/d/1XLzMnGXZlWyLJiolYbK_iNRgCt1ms2-gQbRoq3vPiTY/edit?gid=391151018#gid=391151018)).

I did a sweeping test of the parameter space and landed on these numbers. I tried increasing/decreasing each of the weights (individually) and found that the results were worse (or at least the same). I'm sure that the parameters could be tuned more optimally, but I am also worried that more time spent parameter tuning risks over-fitting the weights to the handful of benchmarks I am using.   

 Is there anything more I can do to justify my weight choices other than pointing to experimental results? I anticipate this could end up being a hurdle when trying to upstream.

I talk a bit more about this in some other comments below too https://github.com/roberttoyonaga/graal/pull/3#discussion_r2685098986

zakkak replied:
> Is there anything more I can do to justify my weight choices other than pointing to experimental results? I anticipate this could end up being a hurdle when trying to upstream.

I am afraid there is not much to do regarding this. Let's see what the GraalVM team thinks about it when the time comes.

FWIW in [An Optimization-Driven Incremental Inline
Substitution Algorithm for Just-in-Time Compilers](http://aleksandar-prokopec.com/resources/docs/prio-inliner-final.pdf) they mention:

> Note that, in our implementation, we calculate No(n) only
for the simplest optimizations, such as constant folding, strength
reduction or branch pruning, and give them all equal weight.
Refining this could yield more precise inlining decisions, but
we did not investigate this in the current work.

galderz commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/NonTrivialInliningGraphDecoder.java:
```diff
@@ -0,0 +1,187 @@
+        double calleeCost = (currentSize - calleeInfo.sizeBeforeInlining);
+
+        double offset = 1.0;
+        double bc = (offset + inlineScope.benefit) * Math.pow(root.compilationInfo.callsites.get(), 2) / calleeCost;
```
What is the origin of this formula and associated parameter values? Same also for threshold. Did you obtain it from the paper or did you come with it yourself?

roberttoyonaga replied:
This formula is based on the "adaptive threshold function" detailed in _[An Optimization-Driven Incremental Inline
Substitution Algorithm for Just-in-Time Compilers]( http://aleksandar-prokopec.com/resources/docs/prio-inliner-final.pdf)_  
<img width="928" height="227" alt="image" src="https://github.com/user-attachments/assets/8919c986-c990-48f2-909e-e2b93c975e81" />

However, I made some slight modifications:
- Instead of using call frequency (no profile data avaiable), I use the square of the root method's callsite count
  - Callsite count is not a perfect proxy for call frequency, but emprirically I found that using it results in significant improvement.  
- I do not include root size in the threshold's exponential term. Including root size here would force all callees to use a shared budget. If the budget is shared, then we should prioritize which callees to evaluate first based on their benefit|cost. I did this in an earlier implementation, but it increased the code complexity and build times a huge amount, so I rolled it back.

zakkak replied:
> Instead of using call frequency (no profile data avaiable), I use the square of the root method's callsite count 

Using the square of the callsite count seems quite aggressive to me.

Using the square of the number of callsites as the call frequency intuitively translates to:
* If a method has a single callsite we expect that this callsite is reached only once
* If a method has two callsites we expect that this callsite is reached four times
* If a method has three callsites we expect that this callsite is reached nine times
* and so forth

This exponential growth of frequency doesn't seem representative of typical programs. On the contrary, I would expect methods being invoked from a single callsite inside a loop to be invoked more frequently in comparison to say a `log` method that's being invoked 5 times in a method.

> Callsite count is not a perfect proxy for call frequency, but emprirically I found that using it results in significant improvement.

It might be interesting to see how often the callsite count is `>1`, how often `>2` and so forth in the experiments and detect the patterns that contributed to that significant improvement.

Potential future work: It would be nice if the algorithm could detect callsites that are in "hot" loops and add the loop "hotness" as an extra weight to the benefit calculation.

> I do not include root size in the threshold's exponential term. Including root size here would force all callees to use a shared budget. If the budget is shared, then we should prioritize which callees to evaluate first based on their benefit|cost. 

Can you please elaborate on this? 

My understanding is that the goal of the adaptive threshold is to increase as the cumulative cost of the optimizations increases. Ignoring the root size removes that property, making the threshold solely dependent on the inlining cost.

Regarding prioritization, I believe I am missing something because the way I understand it would happen _automatically_ as callees with lower benefit|cost ratio would fail to pass the threshold.

roberttoyonaga replied:
>  Using the square of the callsite count seems quite aggressive to me.
> 
> Using the square of the number of callsites as the call frequency intuitively translates to:
> 
> - If a method has a single callsite we expect that this callsite is reached only once
> - If a method has two callsites we expect that this callsite is reached four times
> - If a method has three callsites we expect that this callsite is reached nine times and so forth
>
> This exponential growth of frequency doesn't seem representative of typical programs. On the contrary, I would expect methods being invoked from a single callsite inside a loop to be invoked more frequently in comparison to say a log method that's being invoked 5 times in a method.

I agree that callsite count is not an ideal proxy for call frequency, but I noticed it had a big impact on throughput, at least with the Quarkus benchmarks.  My goal with squaring the callsite count term is to help inline small but very commonly called methods like with`StringLatin1::charAt` that otherwise would have no canonicalization benefit. When I talked to Christian Haeubl, he also didn't like the idea of using callsite count - so I tried removing it, but noticed performance decreased. That was a few months ago and I've since made some small changes, maybe the perf difference will be less now. I'll try testing this again. And I'll also try the case where the callsite count is still used, but is not squared. 


> It might be interesting to see how often the callsite count is >1, how often >2 and so forth in the experiments and detect the patterns that contributed to that significant improvement.

That would be quite interesting. I can try adding some logging to make simple histograms with the collected callsite count data. 

> Potential future work: It would be nice if the algorithm could detect callsites that are in "hot" loops and add the loop "hotness" as an extra weight to the benefit calculation.

I tried implementing this earlier, but I wasn't able to find a way to predict in the static analysis when loops are hot or cold. Do you know how this could be done? 

> Can you please elaborate on this?
> My understanding is that the goal of the adaptive threshold is to increase as the cumulative cost of the optimizations increases. Ignoring the root size removes that property, making the threshold solely dependent on the inlining cost.

Yes that's true. Another point that makes it "adaptive" is that it is more lenient towards small methods.  But in this implementation each callee is evaluated independently like in the Trivial phase so some of the adaptiveness is lost. 

> Regarding prioritization, I believe I am missing something because the way I understand it would happen automatically as callees with lower benefit|cost ratio would fail to pass the threshold.

Once we introduce benefit into the equation the order in which we evaluate callees begins to matter when we have a shared budget. We need to prioritize who to spend budget on first. Callees evaluated earlier have a better chance at passing the threshold since more budget remains available. If we evaluate callees in a greedy way, we may not get the most optimal ordering. 
For example:
-  say Callee_A has low benefit and high cost, while Callee_B and Callee_C have high benefit and low/moderate cost.
- If Callee_A is evaluated first, it may pass the threshold, but it will consume most of the budget. Then there may not be enough budget to inline Callee_B and Callee_C (which are more desireable to inline)
- However, if we do prioritization, we can inline Callee_B and Callee_C. Then Callee_A will be evaluated last and fail the threshold.

That being said, Hotspot uses a greedy strategy. Roland said that Hotspot should ideally switch to using prioritization in the future though.  But if greedy inlining is good enough for Hotspot, perhaps it can be good enough for us too.

zakkak replied:
>> Potential future work: It would be nice if the algorithm could detect callsites that are in "hot" loops and add the loop "hotness" as an extra weight to the benefit calculation.
>
> I tried implementing this earlier, but I wasn't able to find a way to predict in the static analysis when loops are hot or cold. Do you know how this could be done?

No, but I would start by looking at loop optimization phases (e.g. loop unrolling) to see if there is any information used there that could assist on that front.

I would expect constant propagation to allow you to get the actual iterations of some loops.

> However, if we do prioritization, we can inline Callee_B and Callee_C. Then Callee_A will be evaluated last and fail the threshold.

I see. Thanks for the clarification.

(Repeating what's already said to have it in one _paragraph_)

Essentially the difference we are discussing is between

$${b\over{cost}} >= t_1\cdot2^{cost/(16\cdot T_2)}$$
 
and

$${b\over{cost}} >= t_1\cdot2^{(callerSize + cost)/(16\times T_2)}$$

The former version takes in consideration only the cost, thus making it harder to inline larger callees while ignoring the total size of the caller. This indeed seems to resolve the need for prioritization, as any callee falling under the threshold (which is based solely on benefit vs cost) gets inlined.

The latter version introduces the caller size of the resulting code into the equation. As a result it becomes more and more difficult to inline a callee as more callees get inlined.

The former is expected to yield better (and more consistent) performance as callees with similar $b\over{c}$ values will get inlined regardless the caller size. However, it sets no bound on the total code increase incurred by the inlining, i.e., there is no shared budget as you already mentioned. That may result in significant code area increase in cases where methods happen to have multiple callees with good $b\over{c}$ ratio. Setting such a bound, essentially results in a shared budget and requires prioritization to better utilize that budget. Unfortunately, prioritization seems to be expensive and result in significantly longer build times.

galderz commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java:
```diff
@@ -455,6 +504,8 @@ public void finish(DebugContext debug) {
             try (ProgressReporter.ReporterClosable _ = reporter.printInlining()) {
                 inlineTrivialMethods(debug);
+                inlineNonTrivialMethods(debug);
+                inlineSingleCallsiteMethods(debug);
```
Did you consider/try changing the order of non trivial and single call site method inlining? I'm wondering if single call site inlining would happen first, maybe that opens more opportunities for the non trivial one? Whatever the best order is, it might make sense to add some documentation on why the order is what it is.

Update 1: After reading further it's clear that the current implementation relies on the ordering here due to non trivial inline capture the call site count, which is used by the single call site inline. But I wonder, would it be possible to capture the call site count in say, the trivial inliner and then have more flexibility on the order of the non trivial and single call site inliner?

Update 2: In the slides you explain why the single call site inliner is not happening at the same time as the non trivial inliner. But if the single call site inliner was still a separate phase but happened before the non trivial inliner, what would things look like?

zakkak replied:
+1 on explaining why ordering is important, and why that's the best/correct order to use

roberttoyonaga replied:
Theoretically, my intention was to have the Trivial stage happen first so that the more intensive/slow Non-Trivial stage has less callees to evaluate. The Non-Trivial stage takes the longest to complete by a significant margin because it needs to perform trialing. So the idea is to clear away some quick wins with the Trivial stage before engaging Non-Trivial.  SingleCallsite inlining could theoretically happen at any time, but I ordered it last because I am already recording callsite data during the Non-Trivial stage. 

>I'm wondering if single call site inlining would happen first, maybe that opens more opportunities for the non trivial one?

Yes, this is a really good point. I thought about this as well and experimented with adding more Trivial and Non-Trivial stages after the SingleCallsite stage.     _ex: Trivial, Non-Trivial, SingleCallsite, Trivial, Non-Trivial_
Unfortunately, This did not yield significant improvement, but it did increase the build time. See raw [results](https://docs.google.com/spreadsheets/d/1XLzMnGXZlWyLJiolYbK_iNRgCt1ms2-gQbRoq3vPiTY/edit?gid=1204138055#gid=1204138055). 

> But I wonder, would it be possible to capture the call site count in say, the trivial inliner and then have more flexibility on the order of the non trivial and single call site inliner?  
> ...
> But if the single call site inliner was still a separate phase but happened before the non trivial inliner, what would things look like?

Yes, it would be possible to collect the callsite counts in the Trivial stage, and it's also possible to change the order of the SingleCallsite stage (as long as it remains separate). I haven't yet tried reordering the stages so that the SingleCallsite stage happens earlier (I've only added more stages afterward). This is something I'll look into now.

galderz commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/InliningGraphDecoder.java:
```diff
@@ -0,0 +1,66 @@
+        return ((HostedMethod) method).compilationInfo.getCompilationGraph().getEncodedGraph();
+    }
+
+    @Override
```
Is this override necessary? You're just calling the super impl

roberttoyonaga replied:
Yes you're right, I have removed it now

galderz commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/NonTrivialInliningGraphDecoder.java:
```diff
@@ -0,0 +1,187 @@
+        return ((HostedMethod) method).compilationInfo.getCompilationGraph().getEncodedGraph();
+    }
+
+    @Override
```
Same here, is this override necessary? You're just calling the super impl

roberttoyonaga replied:
Yes you're right. Removed!

galderz reviewed (changes requested):
Fantastic work @roberttoyonaga!! I had a further question on testing.

Have you thought about adding some automated unit/integration tests as part of the PR to make sure the logic carries on working in the future? I'm unsure how feasible that is.

zakkak commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/SimplifyingGraphDecoder.java:
```diff
@@ -198,7 +198,8 @@ protected void handleFixedNode(MethodScope methodScope, LoopScope loopScope, int
             Node canonical = canonicalizeFixedNode(methodScope, loopScope, node);
             if (canonical != node) {
-                handleCanonicalization(loopScope, nodeOrderId, node, canonical);
+                methodScope.benefit++;
```
I think this would be better moved inside `handleCanonicalization` and assign a different benefit per case like you already do for the successor deletions.

roberttoyonaga replied:
I agree. That would be more neat and consistent.
I'm going to think a bit more about whether the other cases require different weights.

zakkak commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/SimplifyingGraphDecoder.java:
```diff
@@ -394,7 +399,9 @@ protected Node handleFloatingNodeBeforeAdd(MethodScope methodScope, LoopScope lo
                      * to add additional usages later on for which we need a node. Therefore, we
                      * just do nothing and leave the node in place.
                      */
+                    methodScope.benefit += 20;
```
It's not clear to me why this has such "big" benefit while doing nothing.
Is it because we know that they are going to be "killed"?

roberttoyonaga replied:
Yes, it's because we know the node will later be removed at the end of the round (if there are no further usages). So it's prioritizing trimming down the graph as much as possible.

zakkak commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java:
```diff
@@ -895,14 +895,14 @@ public static long getTearDownFailureNanos() {
     public static final HostedOptionKey<AccumulatingLocatableMultiOptionValue.Strings> NeverInlineTrivial = new HostedOptionKey<>(AccumulatingLocatableMultiOptionValue.Strings.build());
 
-    @Option(help = "Maximum number of nodes in a method so that it is considered trivial.")//
-    public static final HostedOptionKey<Integer> MaxNodesInTrivialMethod = new HostedOptionKey<>(20);
```
Would it make sense to Deprecate this and use its value to provide an estimation of the corresponding size when it's used?

E.g. 20 nodes could translate to size 60

roberttoyonaga replied:
Yes I think deprecating these hosted options could be better than removing them entirely.  However, I think that if we want to keep the old options we should support both old and new implementations in `substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/InliningUtilities.java` , rather than translate the user-supplied node count to an estimated size. Otherwise we could give an unexpected result to the user, especially if we are doing this translation transparently.
 What do you think? @zakkak

zakkak replied:
I agree that supporting both would be best.

zakkak commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/SubstrateOptions.java:
```diff
@@ -895,14 +895,14 @@ public static long getTearDownFailureNanos() {
     public static final HostedOptionKey<Integer> MaxInvokesInTrivialMethod = new HostedOptionKey<>(1);
 
-    @Option(help = "Maximum number of nodes in a method so that it is considered trivial, if it does not have any invokes.")//
-    public static final HostedOptionKey<Integer> MaxNodesInTrivialLeafMethod = new HostedOptionKey<>(40);
```
Same question as in https://github.com/roberttoyonaga/graal/pull/3/changes#r2693773783

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CalleeInfo.java:
```diff
@@ -0,0 +1,53 @@
+ * {@link HostedMethod}'s {@link com.oracle.svm.hosted.code.CompilationInfo} owns a collection of
+ * {@link CalleeInfo} objects.
+ */
+public class CalleeInfo {
```
Does this need to be `public`?

roberttoyonaga replied:
Good catch, I have changed it to be package-private

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java:
```diff
@@ -752,13 +803,20 @@ private static boolean checkNewlyTrivial(HostedMethod method, StructuredGraph gr
+     * constraints and focus only on methods with a high benefit. If the non-trivial inlining was
+     * responsible for inlining trivial methods as well it would need more flexible cost constraints
+     * to handle trivial methods with no computed benefit. Such looser constraints would allow
+     * larger methods with (proportionally more benefit) to more easily be inlined.
```
Do I understand correctly that the problem with this would be non-acceptable code-size increase? If so, can you please make the clarification in the comment?

roberttoyonaga replied:
Yes, that is correct. I have made the javadoc a bit clearer

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/InliningUtilities.java:
```diff
@@ -41,31 +35,27 @@ public class InliningUtilities {
-        int numOthers = 0;
+        int size = 0;
         for (Node n : graph.getNodes()) {
-            if (n instanceof StartNode || n instanceof ParameterNode || n instanceof FullInfopointNode || n instanceof ValueProxy || n instanceof ValueAnchorNode || n instanceof FrameState) {
-                continue;
-            }
```
What's the reason of going over these? Why not ignore them like previously? I see that their size is zero.

roberttoyonaga replied:
Good point, we can continue to skip over most of them. Upon further inspection, It looks like `FrameState` and `ParameterNode` can have non-zero size, so we shouldn't skip over those ones.  

Upon testing, if I skip over those while building the Quarkus getting-started quickstart the code area and file size increases:
 Code area: 20.95MiB →  21.27MiB 
 Binary size: 47.69MiB →  48.00MiB   

I've partially applied your suggestion but still evaluate `FrameState` and `ParameterNode`

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/NonTrivialInliningGraphDecoder.java:
```diff
@@ -0,0 +1,187 @@
+         * Stash the graph size in the CalleeInfo. Recursion (due to multiple callsites at different
+         * depths) should not be a problem since we only go one level deep per round.
+         */
+        root.compilationInfo.callees.get(callee).sizeBeforeInlining = NodeCostUtil.computeGraphSize(graph);
```
I suggest using `get` instead of `containsKey` and checking against `null`, this way you avoid getting the node twice, once to see if it exists and another to get its actual value.

```suggestion
        // If needed, create a CalleeInfo for the current callee
        CalleeInfo calleeInfo = root.compilationInfo.callees.get(callee);
        if (calleeInfo == null) {
            // If this callee is inlined, this CalleeInfo will not survive beyond the current round.
            calleeInfo = new CalleeInfo(callee, round);
            root.compilationInfo.callees.put(callee, calleeInfo);
        }
        /*
         * Stash the graph size in the CalleeInfo. Recursion (due to multiple callsites at different
         * depths) should not be a problem since we only go one level deep per round.
         */
        calleeInfo.sizeBeforeInlining = NodeCostUtil.computeGraphSize(graph);
```

roberttoyonaga replied:
Good suggestion! I have adopted it

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/NonTrivialInliningGraphDecoder.java:
```diff
@@ -0,0 +1,187 @@
+        HostedMethod root = (HostedMethod) callerScope.method;
+
+        if (!canInline(inlineScope, root, callee)) {
+            // This block is essentially the same as InlineBeforeAnalysisGraphDecoder#finishInlining
```
Did you try extracting it in a method to avoid duplication?

roberttoyonaga replied:
That's a great point. It is indeed possible to move that chunk into `PEGraphDecoder`. Applied!

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompilationInfo.java:
```diff
@@ -45,6 +47,12 @@
 
 public class CompilationInfo {
+    public int sizeLastRound;
+    public AtomicLong callsites = new AtomicLong();
```
Isn't an Integer enough?

roberttoyonaga replied:
The inlining decoding for each root method happens in parallel on as many threads as are available in the compiler threadpool. So all the root methods are decoded concurrently. Please see where `inlineNonTrivialMethods` calls [`runOnExecutor`](https://github.com/roberttoyonaga/graal/blob/694e3157dbc6707fe5ff12b8999f15ef958905d8/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java#L860)

Unlike `MethodScope`s (which are owned exclusively by root's respective decoder), each method's `CompilationInfo` can be accessed from other threads decoding other roots concurrently. 

The other `CompilationInfo` fields like [`sizeLastRound`](https://github.com/roberttoyonaga/graal/pull/3/files#diff-e75076ac7a215723c7e97570760f801c17ddfbac7c1cdf4ab4c80a700d0382afR50) and `callees` do not need synchronization because I only allow them to be accessed by the thread performing that respective root's decoding. However,  `CompilationInfo#callsites` can be written from any other decoding thread.  Whenever a callsite of a method is discovered (by any thread) it must be recorded in that method's `CompilationInfo`

zakkak replied:
Sorry for not being more specific. I meant isn't an `AtomicInteger` (in contrast to an `AtomicLong`) enough?

roberttoyonaga replied:
Oh yes I see. `AtomicInteger` should indeed be enough. I've changed it now

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java:
```diff
@@ -920,6 +1165,116 @@ private boolean makeInlineDecision(HostedMethod method, HostedMethod callee) {
+            s.newCallees.put(callee, newCount);
+        } else {
+            s.newCallees.put(callee, 1);
+        }
```
Avoid calculating the hash twice on the slow path. 

```suggestion
        Integer count = s.newCallees.get(callee);
        if (count == null) {
            s.newCallees.put(callee, 1);
        } else {
            s.newCallees.put(callee, count + 1);
        }
```

roberttoyonaga replied:
Good point. Applied!

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java:
```diff
@@ -920,6 +1165,116 @@ private boolean makeInlineDecision(HostedMethod method, HostedMethod callee) {
+    }
+
+    private static void updateCallsiteCountRecords(PEGraphDecoder.PEMethodScope targetScope, HostedMethod callee) {
+        NonTrivialInliningGraphDecoder.NonTrivialInliningMethodScope s = (NonTrivialInliningGraphDecoder.NonTrivialInliningMethodScope) targetScope;
```
I would set the type in the parameters and let the caller handle the casting (and the type check).

```suggestion
    private static void updateCallsiteCountRecords(NonTrivialInliningGraphDecoder.NonTrivialInliningMethodScope targetScope, HostedMethod callee) {
```

roberttoyonaga replied:
Good point. Applied!

zakkak reviewed (commented):
Great work @roberttoyonaga !!!

I have added some specific inline comments and would like to propose the following high level comments in preparation of an upstream PR:

1. IMHO the PR should better be split in (preferably) three smaller PRs. One for the `trivialInliner` changes. Another for the `singleCallsiteInliner` (this might not be trivial as it requires moving the callsite counting in the `trivialInliner`, so better discuss it with others as well before puttig the effort). And a third for the `nonTrivialInliner`. I believe the trivial inliner changes along with the singlecallsite inliner will be easier/faster to get accepted.
2. Ideally the PR should be accompanied with an evaluation of the impact of each inliner separately. The evaluation should mention at least the following:
  * Impact on image size (code area)
  * Impact on run-time performance
  * Impact on build-time heap usage (RSS?)
  * Impact on build execution time

roberttoyonaga committed:
Apply review feedback: refactoring, comments, small fixes

roberttoyonaga commented:
Thank you Galder and Foivos very much for your reviews!!

I will break this PR into 3 smaller PRs. First for the Trivial inliner changes, second the SingleCallsite stage, and finally the Non-Trivial stage.  I think it makes sense to try to upstream them in that order. 

I've pushed a commit that applies most of the refactoring and smaller feedback.  However, there are still a few things that are unresolved and I need to think about some more. 


----

@galderz 
> Have you thought about adding some automated unit/integration tests as part of the PR to make sure the logic carries on working in the future? I'm unsure how feasible that is.

 Yes, I had thought about adding tests. However, the problem is that I am not sure how to check whether test methods have been inlined correctly without building the test  executable with debug info and checking the assembly.  Maybe there is an automated way to do that, but the native-unittests are a bit limited. I will ask the GraalVM team and see if they can suggest a good approach.

Perhaps there is a way to directly unit test the decoder classes. Something like feeding a dummy graph to the decoders and verifying the result. I'm not sure this is possible, but I will brainstorm this some more. 

----

@zakkak 
>I have added some specific inline comments and would like to propose the following high level comments in preparation of an upstream PR:
>1. IMHO the PR should better be split in (preferably) three smaller PRs. One for the trivialInliner changes. Another for the singleCallsiteInliner (this might not be trivial as it requires moving the callsite counting in the trivialInliner, so better discuss it with others as well before puttig the effort). And a third for the nonTrivialInliner. I believe the trivial inliner changes along with the singlecallsite inliner will be easier/faster to get accepted.
>2. Ideally the PR should be accompanied with an evaluation of the impact of each inliner separately. The evaluation should mention at least the following:
> - Impact on image size (code area)
> - Impact on run-time performance
> - Impact on build-time heap usage (RSS?)
> - Impact on build execution time


I think this a good idea and would probably help get this work integrated sooner. Moving the callsite counting out of the Non-Trivial stage wouldn't be too hard. There are 2 options:
 - As you said, do the callsite counting while we process the Trivial stage
 - Do an initial round in the SingleCallsite stage for counting callsites as we pass through the root graphs. I've already experiemented with this approach before.  It doesn't extend build time too much. 

I've already been tracking impact on image size, runtime perf, and build time, but I have not considered tracking build time heap usage. I think that's a good idea and will start tracking that metric too.

zakkak commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/SimplifyingGraphDecoder.java:
```diff
@@ -67,6 +67,18 @@
+     * doesn't change the overall structure of the graph. These nodes may not even get removed
+     * during the canonicalizer phase if they are used again elsewhere. Fixed node removal is
+     * weighted more becuase it changes graph structure. Removal of conditional blocks is weighted
+     * the most becuase it reduces branching.
```
```suggestion
     * weighted more because it changes graph structure. Removal of conditional blocks is weighted
     * the most because it reduces branching.
```

roberttoyonaga replied:
applied!

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/NonTrivialInliningGraphDecoder.java:
```diff
@@ -0,0 +1,170 @@
+ * made. This allows expected cost and benefit to be computed. If inlining is denied, then the nodes
+ * moved into the caller are rolled back similarly to what is done in the
+ * {@link com.oracle.graal.pointsto.phases.InlineBeforeAnalysisGraphDecoder}.
+ */
```
Please document that this is implementing a variation of [An Optimization-Driven Incremental Inline
Substitution Algorithm for Just-in-Time Compilers](http://aleksandar-prokopec.com/resources/docs/prio-inliner-final.pdf) explicitly stating any deviations from it.

roberttoyonaga replied:
Done!

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java:
```diff
@@ -771,17 +829,139 @@ protected void inlineTrivialMethods(DebugContext debug) throws InterruptedExcept
+        if (inliningRound == 1) {
+            /*
+             * Skip roots that are themselves single callsite methods. Otherwise, it's possible that
+             * the both the root and some of its callees are both single callsite methods. In such
```
```suggestion
             * the root and some of its callees are both single callsite methods. In such
```

roberttoyonaga replied:
Applied!

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/NonTrivialInliningGraphDecoder.java:
```diff
@@ -0,0 +1,170 @@
+        double offset = 1.0;
+        double bc = (offset + inlineScope.benefit) * Math.pow(root.compilationInfo.callsites.get(), 2) / calleeCost;
+        double t1 = 5;
+        double t2 = 1;
```
Please make these constant fields as well.

You will also need to kind of justify why they were chosen. The paper uses `t1=0.005` and `t2=120` (without any justification unfortunately), so how did you come up with a different set? (one reason is probably the fact that you use different metrics both for the benefit and the cost)

roberttoyonaga replied:
Yes I also chose these constants empirically. I tried using the same constants as in the paper but they produced really bad results because the threshold function is not exactly the same. 

I'll make these constant fields and add some explanation about why they were chosen.

zakkak commented on compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/SimplifyingGraphDecoder.java:
```diff
@@ -67,6 +67,18 @@
+    private static final int FLOATING_REMOVAL_WEIGHT = 20;
+    private static final int FIXED_SUCCESSOR_REMOVAL_WEIGHT = 30;
+    private static final int FIXED_REMOVAL_WEIGHT = 35;
+    private static final int CONDITIONAL_REMOVAL_WEIGHT = 55;
```
The paper states that they treat all optimizations equally, so the weighted approach seems like a big deviation from the original algorithm. Please make sure to discuss this in https://github.com/roberttoyonaga/graal/pull/3/changes#r2697757729

I believe that these constants should be moved under `NonTrivialInliningGraphDecoder`, but I see that's in a completely different package/module, so not sure if it's possible.

roberttoyonaga replied:
Hmmm I agree that would be a lot nicer.  I made `NonTrivialInliningGraphDecoder`  and those fields public, but got this Graal build error :  `package com.oracle.svm.hosted.code does not exist`

I think referencing substratevm from the compiler suite violates the build ordering. The compiler suite probably needs to be built first. 

One argument for those constant staying in `SimplifyingGraphdecoder` is that we are already awarding benefit there, so it could make sense to keep all the code related to awarding benefit in one place.

zakkak commented:
> I will break this PR into 3 smaller PRs. First for the Trivial inliner changes, second the SingleCallsite stage, and finally the Non-Trivial stage. I think it makes sense to try to upstream them in that order.

May I make another suggestion?

Can you put the `nonTrivialInliner` behind a flag? I suspect that even if accepted upstream it would be nice to be able to disable (if it turns up causing issues), or even have it off by default and add is a [future default](https://github.com/oracle/graal/blob/5e7877a3dd5089beb2523068c9d3b89ea0d3c5d4/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/FutureDefaultsOptions.java) for testing.

roberttoyonaga committed:
add hosted options for inline stages, use atomic int, comments

roberttoyonaga committed:
javadoc and comments

roberttoyonaga commented:
> May I make another suggestion?
> 
> Can you put the nonTrivialInliner behind a flag? I suspect that even if accepted upstream it would be nice to be able to disable (if it turns up causing issues), or even have it off by default and add is a [future default](https://github.com/oracle/graal/blob/5e7877a3dd5089beb2523068c9d3b89ea0d3c5d4/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/FutureDefaultsOptions.java) for testing.

Good suggestion! I have added a hosted option flag for this.  I also added an option for the SingleCallsite inlining for completion's sake.

zakkak commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java:
```diff
@@ -455,6 +506,8 @@ public void finish(DebugContext debug) {
             }
             try (ProgressReporter.ReporterClosable _ = reporter.printInlining()) {
                 inlineTrivialMethods(debug);
+                inlineNonTrivialMethods(debug);
+                inlineSingleCallsiteMethods(debug);
```
I would expect something like

```suggestion
                if (SubstrateOptions.AOTNonTrivialInline.getValue()) {
                    inlineNonTrivialMethods(debug);
                }
                if (SubstrateOptions.AOTSingleCallsiteInline.getValue()) {
                    inlineSingleCallsiteMethods(debug);
                }
```

Why is this not the case for the existing `inlineTrivialMethods(debug);` call?

roberttoyonaga replied:
I was wondering that too. The reason that `optionAOTTrivialInline` needs to be read  in `makeInlineDecision` is because there are some methods that should always be inlined (due to `AlwaysInline` annotations) - regardless of whether the user wants to run the inliner. See [`callee.shouldBeInlined()`](https://github.com/roberttoyonaga/graal/blob/233b9077b1a89b4b21eb2f274aa08a1f40bc846f/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/code/CompileQueue.java#L1162) in `makeInlineDecision`.  So we always need to run the Trivial stage and check each callee.

As for Non-Trivial and SingeCallsite:
Originally, I thought that inlining in previous stages may add new callees marked `AlwaysInline`  to each root method's scope, so we should always run the Non-Trivial and SingleCallsite stages. But after thinking about it more, I realize that is wrong.  Each of the stages runs until no more inlining can be done, and a steady state is reached. This means it's impossible for there to be un-inlined callees marked `AlwaysInline` in the scopes of root methods at the end of the previous stage. I have added a new commit adopting your suggestion. Thanks for catching this!

roberttoyonaga committed:
possibly don't run nonTrivial and singlecallsite

roberttoyonaga committed:
support both old and new Trivial inlining hosted options

roberttoyonaga commented:
I've been working on splitting this into multiple smaller PRs (one per stage). I've opened PRs for the [Trivial](https://github.com/roberttoyonaga/graal/pull/4) and [SingleCallsite](https://github.com/roberttoyonaga/graal/pull/5) stages. The functionality is basically the same. I've just factored out the relevant code.   

In each PR description I've also included benchmark results of each of the stages independent of the other stages. The Trivial stage PR does not have much of an impact. So I think we should lead with the SingleCallsite PR when contributing this upstream.

I'm still investigating the feasibility of unit tests. There are currently no tests for the existing inliner. I do not think it will be possible without building a testing harness from the ground up that can hook into the Native Image builder.  The few tests that exist for the Graal JIT inliner are not useful because they do not have access to the hosted context (such as `HostedMethod.compilationInfo`).  The best course of action is probably to open the upstream PR and ask Oracle for their suggestions on how to test before investing a lot of time into developing anything ourselves.

zakkak commented:
Great work @roberttoyonaga. I am adding comments on the corresponding PRs.

> The best course of action is probably to open the upstream PR and ask Oracle for their suggestions on how to test before investing a lot of time into developing anything ourselves.

I agree
